File: 53-maximum-subarray/maximum-subarray.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int max_sum = INT_MIN;
        int current_sum = 0;

        for (int num : nums) {
            current_sum += num;
            if (current_sum > max_sum) {
                max_sum = current_sum;
            }
            if (current_sum < 0) {
                current_sum = 0;
            }
        }

        return max_sum;
    }
};

File: 53-maximum-subarray/maximum-subarray.rs
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {

        let mut max = std::i32::MIN;
        let mut sum = 0;

        for &num in nums.iter() {
            sum += num;
            if sum > max {
                max = sum;
            }
            if sum < 0 {
                sum = 0;
            }
        }

        max
    }
}